### DISCUSSION TITLE
"Unable to write data to the transport connection" exception

### DISCUSSION THREAD

**Discussion by teSill** [jchristn/WatsonTcp#244] [Q&A]
Hey,

I've recently been getting a few exceptions that I can't quite figure out how to prevent or handle.

The first one is this:
` Unable to write data to the transport connection: Connection reset by peer.:    at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
at WatsonTcp.WatsonTcpServer.SendHeaders(ClientMetadata client, WatsonMessage msg)
at WatsonTcp.WatsonTcpServer.SendInternal(ClientMetadata client, WatsonMessage msg, Int64 contentLength, Stream stream)`

And the second one that follows shortly after is:
`Unable to write data to the transport connection: Broken pipe.:    at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
at WatsonTcp.WatsonTcpServer.SendHeaders(ClientMetadata client, WatsonMessage msg)
at WatsonTcp.WatsonTcpServer.SendInternal(ClientMetadata client, WatsonMessage msg, Int64 contentLength, Stream stream)`

It's happened twice now after a specific event in the app, where the server sends a bunch of data to dozens of clients (which is way more than what is normal for the application). Not sure if it directly relates to that though. 

Any insight on what the exception really means, what might be causing it and how to prevent it?

Thank you!

**Comment by teSill** [comment]
Another slightly different exception that seems to sometimes follow the ` Unable to write data to the transport connection` exception:

`Cannot access a disposed object.
Object name: 'System.Net.Sockets.NetworkStream'.:    at System.Net.Sockets.NetworkStream.Write(Byte[] buffer, Int32 offset, Int32 size)
at WatsonTcp.WatsonTcpServer.SendHeaders(ClientMetadata client, WatsonMessage msg)
at WatsonTcp.WatsonTcpServer.SendInternal(ClientMetadata client, WatsonMessage msg, Int64 contentLength, Stream stream)`

I've tried catching the exception from WatsonTcpServer.Send but to no avail. The errors are unfortunately causing the app to become unresponsive to a lot of actions and connections, requiring a swift restart when it occurs.

**Reply by jchristn** [reply]
Hi @teSill could you produce a code sample (perhaps from Test.Client and Test.Server projects) that highlight this behavior?  Usually these exceptions point to the network peer having disconnected and the library not doing the right thing (i.e. accessing a disposed object) or not exiting/ending tasks correctly.

**Reply by teSill** [reply]
I'm not sure what's causing the issue, I haven't been able to reproduce it myself. Note that I was running on a pretty old version of WatsonTcp when I made this post but have updated to the latest since then. I haven't seen the exceptions since, but a new one came up last night: `Null header data indicates peer disconnected`. Not sure if it's an issue in my disconnection handling or the library's but it seems that after receiving this exception the server was unable to process anymore connections/logins at all.

Edit: have been seeing the exception occasionally now but it hasn't caused any visible problems. Guess the issue before wasn't directly related to that exception.
